controller/src/beerocks/master/son_master_thread.cpp

```diff
@@ -1186,16 +1191,162 @@ bool master_thread::handle_cmdu_1905_higher_layer_data_message(Socket *sd,
     return son_actions::send_cmdu_to_agent(sd, cmdu_tx);
 }
 
+bool master_thread::handle_cmdu_1905_topology_notification(Socket *sd,
+                                                           ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received TOPOLOGY_NOTIFICATION_MESSAGE, mid=" << std::hex << int(mid);
+
+    std::shared_ptr<wfa_map::tlvClientAssociationEvent> client_association_event_tlv = nullptr;
+    std::shared_ptr<beerocks_message::tlvVsClientAssociationEvent> vs_tlv            = nullptr;
+
+    int type;
+    while ((type = cmdu_rx.getNextTlvType()) != int(ieee1905_1::eTlvType::TLV_END_OF_MESSAGE)) {
+        switch (type) {
+        case int(wfa_map::eTlvTypeMap::TLV_CLIENT_ASSOCIATION_EVENT):
+            LOG(DEBUG) << "Found TLV_CLIENT_ASSOCIATION_EVENT TLV";
+            client_association_event_tlv = cmdu_rx.addClass<wfa_map::tlvClientAssociationEvent>();
+            if (!client_association_event_tlv) {
+                LOG(ERROR) << "addClass wfa_map::tlvClientAssociationEvent failed";
+                return false;
+            }
+            break;
+        case int(ieee1905_1::eTlvType::TLV_VENDOR_SPECIFIC):
+            LOG(DEBUG) << "Found TLV_VENDOR_SPECIFIC TLV";
+
+            if (!message_com::parse_intel_vs_message(cmdu_rx)) {
+                LOG(DEBUG) << "Failed to parse intel vs message (not Intel?)";
+                continue;
+            }
+
+            vs_tlv = cmdu_rx.addClass<beerocks_message::tlvVsClientAssociationEvent>();
+            if (!vs_tlv) {
+                LOG(ERROR) << "addClass wfa_map::tlvVsClientAssociationEvent failed";
+                return false;
+            }
+            break;
+        default:
+            LOG(ERROR) << "Unexpected TLV. Ignore CMDU.";
+            return false;
+        }
+    }
+
+    if (!client_association_event_tlv) {
+        LOG(ERROR) << "wfa_map::tlvClientAssociationEvent not found";
+        return false;
+    }
+
+    auto &client_mac    = client_association_event_tlv->client_mac();
+    auto client_mac_str = network_utils::mac_to_string(client_mac);
+
+    auto &bssid    = client_association_event_tlv->bssid();
+    auto bssid_str = network_utils::mac_to_string(bssid);
+
+    auto &association_event = client_association_event_tlv->association_event();
+
+    LOG(INFO) << "client "
+              << (association_event == wfa_map::tlvClientAssociationEvent::CLIENT_HAS_JOINED_THE_BSS
```

Comment:
(nitpick) better move this to a bool variable.

```cpp
bool client_connected = association_event == wfa_map::tlvClientAssociationEvent::CLIENT_HAS_JOINED_THE_BSS;
```

Reply:
Good idea, it makes the log print to look nicer.